app/grandchallenge/notifications/migrations/0008_migrate_forum_follows_and_notifications.py

```diff
@@ -0,0 +1,115 @@
+from django.core.exceptions import ObjectDoesNotExist
+from django.db import migrations
+from django.db.models import Q
+
+from grandchallenge.discussion_forums.models import (
+    get_matching_forum,
+    get_matching_topic,
+)
+
+
+def migrate_forum_and_topic_follows_and_notifications(  # noqa C901
+    apps, schema_editor
+):
+    Notification = apps.get_model(  # noqa: N806
+        "notifications", "Notification"
+    )
+    Follow = apps.get_model("actstream", "Follow")  # noqa: N806
+    Forum = apps.get_model("discussion_forums", "Forum")  # noqa: N806
+    ForumTopic = apps.get_model(  # noqa: N806
+        "discussion_forums", "ForumTopic"
+    )
+    MachinaForum = apps.get_model("forum", "Forum")  # noqa: N806
+    MachinaTopic = apps.get_model("forum_conversation", "Topic")  # noqa: N806
+    ContentType = apps.get_model("contenttypes", "ContentType")  # noqa: N806
+
+    new_topic_ct = ContentType.objects.get_for_model(ForumTopic)
+    new_forum_ct = ContentType.objects.get_for_model(Forum)
+    old_forum_ct = ContentType.objects.get_for_model(MachinaForum)
+    old_forumtopic_ct = ContentType.objects.get_for_model(MachinaTopic)
+
+    # delete old notifications with outdated references
+    _, deleted_count = Notification.objects.filter(
+        Q(target_content_type__in=[old_forum_ct, old_forumtopic_ct])
+        | Q(
+            action_object_content_type__in=[
+                old_forum_ct,
+                old_forumtopic_ct,
+            ]
+        )
+    ).delete()
+    print(f"Successfully deleted {deleted_count} notifications.")
+
+    follows_to_update = []
+    n_updated = 0
+    batch_size = 1000
+
+    for follow in (
+        Follow.objects.filter(
+            content_type__in=[old_forum_ct, old_forumtopic_ct]
+        )
+        .only("object_id", "content_type")
+        .iterator(chunk_size=batch_size)
+    ):
+        if follow.content_type == old_forum_ct:
+            try:
+                forum = get_matching_forum(
+                    old_forum_id=follow.object_id,
+                    old_forum_model=MachinaForum,
+                )
+            except ObjectDoesNotExist:
+                continue
+            follow.object_id = forum.pk
+            follow.content_type = new_forum_ct
+        else:
+            try:
+                topic = get_matching_topic(
+                    old_topic_id=follow.object_id,
+                    old_topic_model=MachinaTopic,
+                    new_topic_model=ForumTopic,
+                    old_forum_model=MachinaForum,
+                )
+            except ObjectDoesNotExist:
+                continue
+            follow.object_id = topic.pk
+            follow.content_type = new_topic_ct
+
+        follows_to_update.append(follow)
+
+        if len(follows_to_update) >= batch_size:
+            Follow.objects.bulk_update(
+                follows_to_update,
+                fields=["content_type", "object_id"],
+            )
```

Comment:
In test run 1001 just now, I realized that after we migrate the follows (i.e. updating the content type and object reference of existing follows), new follows of the old content type can be added for the same object id and user (i.e. when a user replies to a topic they were subscribed to). Migrating those in a subsequent run of the migration will fail because the corresponding new follow already exists. The management command and the migration need to be updated to check for whether a follow already exists before adding it to the list to be updated. That makes this migration a bit more complex still but I don't think we have another option.